onnxruntime/core/framework/mem_pattern_planner.h

```diff
@@ -121,8 +128,11 @@ class MemPatternPlanner {
 
     blocks_.insert(best_fit_it, (static_cast<int>(allocs_.size()) - 1));
   }
+#endif
 
   void TraceAllocation(int ml_value_idx, size_t size) {
+    ORT_ENFORCE(!using_counters_);
```

Comment:
> ORT_ENFORCE(!using_counters_); [](start = 3, length = 31)

I like this but unfortunately it will fail in a training scenario where we have graphs that haves symbolic shapes which cannot be resolved ahead of graph execution in this case today we allocate the tensors dynamically and while doing so we trace them at the same time so after the first step we get a pattern for its input that we cache and use in subsequent run if the input shape is the same, honestly it is suboptimal and I will improve it BUT when we trace we use this function, please see the usage of *planner_* in execution_frame.cc, specifically in TraceAllocate. 

May be you can remove this assert and I will change the usage from TraceAllocation(int, size) to TraceAllocate(int, ProgramCounter, size_t) or you can change to if you want to.